include/silo/query_engine/query_result.h

```diff
@@ -17,13 +19,85 @@ struct QueryResultEntry {
    std::map<std::string, common::JsonValueType> fields;
 };
 
-struct QueryResult {
-   std::vector<QueryResultEntry> query_result;
+class QueryResult {
+   // Temporary chunk of the query result in the case of streaming, or
+   // the whole result in the case of eager query evaluation.
+   std::vector<QueryResultEntry> query_result_chunk_;
+   // Receives the cleared `query_result_chunk_` and fills in the next
+   // batch of entries of the result set; it adds no entries iff the
+   // end of the result set has been reached.
+   std::function<void(std::vector<QueryResultEntry>&)> get_chunk_;
+   // Iterating through query_result_chunk_.
+   size_t i_;
+   // Used for safety checks when asking to get all values via
+   // `entries` or `entriesMut`.
+   bool is_materialized_;
+
+  protected:
+   QueryResult(
+      std::vector<QueryResultEntry>&& query_result_chunk,
+      std::function<void(std::vector<QueryResultEntry>&)>&& get_chunk,
+      bool is_materialized
+   )
+       : query_result_chunk_(std::move(query_result_chunk)),
+         get_chunk_(std::move(get_chunk)),
+         i_(0),
+         is_materialized_(is_materialized) {}
+
+  public:
+   /// For eager query evaluation.
+   static QueryResult fromVector(std::vector<QueryResultEntry>&& query_result);
+   /// For streaming query evaluation.
+   static QueryResult fromGenerator(std::function<void(std::vector<QueryResultEntry>&)>&& get_chunk
+   );
+   /// The empty result.
+   QueryResult()
+       : get_chunk_([](std::vector<QueryResultEntry>& /*query_result_chunk*/) {}),
+         i_(0),
+         is_materialized_(true) {}
+
+   // Moves
+   QueryResult(QueryResult&& other) noexcept
+       : query_result_chunk_(std::move(other.query_result_chunk_)),
+         get_chunk_(std::move(other.get_chunk_)),
+         i_(other.i_),
+         is_materialized_(other.is_materialized_){};
```

Comment:
Let's please also move this to the cpp file

Reply:
As discussed I leave it as is (with my current rule for moving implementation when there's something within the body block), but will see if I get input from IRC/disassembly/web on benefits/disadvantage.